core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,387 @@
+            } else {
+                methodFields = structInfo.fields();
+            }
+            visited.put(signature, currentPath);
+        } else {
+            methodFields = methodInfo.parameters();
+        }
+
+        generateProperties(methodFields, visited, currentPath, root);
+        return root;
+    }
+
+    /**
+     * Generate the JSON Schema for the given {@link FieldInfo} and add it to the given {@link ObjectNode}
+     * and add required fields to the {@link ArrayNode}.
+     *
+     * @param field field to generate schema for
+     * @param visited map of visited types and their paths
+     * @param path current path in tree traversal of fields
+     * @param root the root to add schema properties
+     * @param required the array node to add required field names
+     */
+    private void generateField(FieldInfo field, Map<String, String> visited, String path, ObjectNode root,
+                               ArrayNode required) {
+        final ObjectNode fieldNode = mapper.createObjectNode();
+        final String fieldTypeSignature = field.typeSignature().signature();
+
+        if (visited.containsKey(fieldTypeSignature)) {
+            // If field is already visited, add a reference to the field instead of iterating its children.
+            final String pathName = visited.get(fieldTypeSignature);
+            fieldNode.put("$ref", pathName);
+        } else {
+            final String schemaType = getSchemaType(field.typeSignature());
+
+            // Field is not visited, create a new type definition for it.
+            fieldNode.put("type", schemaType);
+            fieldNode.put("description", field.descriptionInfo().docString());
+
+            // Fill required fields for the current object.
+            if (field.requirement() == FieldRequirement.REQUIRED) {
+                required.add(field.name());
+            }
+
+            if (field.typeSignature().type() == TypeSignatureType.ENUM) {
+                fieldNode.set("enum", getEnumType(field.typeSignature()));
+            }
+
+            final String currentPath;
+            if (field.name().isEmpty()) {
+                currentPath = path;
+            } else {
+                currentPath = path + '/' + field.name();
+            }
+
+            // Only Struct types map to custom objects to we need reference to those structs.
+            // Having references to primitives do not make sense.
+            if (MEMORIZED_JSON_TYPES.contains(schemaType)) {
+                visited.put(fieldTypeSignature, currentPath);
+            }
+
+            // Based on field type, we need to call the appropriate method to generate the schema.
+            // For example maps have `additionalProperties` field, arrays have `items` field and structs
+            // have `properties` field.
+            if (field.typeSignature().type() == TypeSignatureType.MAP) {
+                generateMapFields(fieldNode, field, visited, currentPath);
```

Comment:
Shouldn't we also add
```
  "type": "object",
  "properties": {},
```
to the root?

Reply:
If the root is a `Struct`, then it is added inside `generateStructFields`.

But for maps, there are no `properties` because those properties have names. Instead we use `additionalProperties` so we have a generic mapping without caring about what the property names are.

type object is added in Line 157.